### PR TITLE
Fix nearest_neighbour_index for edge case where requested point is float and bounds are int

### LINKED ISSUE #2969: Very specific error with extract
Extracting using a constraint on a scalar coordinate with bounds, where the value of the constraint is a non-integer outside the bounds causes an error:
```python
import iris
from iris.cube import Cube
from iris.coords import DimCoord
x = DimCoord([1], bounds=[0,2], long_name='x')
cube = Cube(1, aux_coords_and_dims=[(x,())])
con = iris.Constraint(x=2.5)
cube.extract(con)
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../iris/cube.py", line 2044, in extract
    return constraint.extract(self)
  File ".../iris/_constraints.py", line 151, in extract
    resultant_CIM = self._CIM_extract(cube)
  File ".../iris/_constraints.py", line 175, in _CIM_extract
    resultant_CIM = resultant_CIM & coord_constraint.extract(cube)
  File ".../iris/_constraints.py", line 273, in extract
    i = coord.nearest_neighbour_index(self._coord_thing)
  File ".../iris/coords.py", line 1189, in nearest_neighbour_index
    result_index = np.where(inside_cells)[0][0]
IndexError: index out of bounds
```

### PR DESCRIPTION
**Pull request by rcomer**
### Description

Fixes #2969.  The example given there now returns `None` as expected.

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Review by wjbenfold — APPROVED** [review]
Note to future me: the original fails because it casts the `float` result of `max` back to `int` in

https://github.com/SciTools/iris/blob/7d26f4d6dbb8b2049a4424881a62dba61a69d025/lib/iris/coords.py#L2226

File: docs/src/whatsnew/latest.rst [hunk 1]
```diff
@@ -62,6 +62,10 @@ This document explains the changes made to Iris for this release
    one cell's bounds align with the requested maximum and negative minimum, fixing
    :issue:`4221`. (:pull:`4278`)
 
+#. `@rcomer`_ fixed :meth:`~iris.coords.Coord.nearest_neighbour_index` for edge
```
Comment (wjbenfold) [inline docs/src/whatsnew/latest.rst:65]: Please rebase before merge

Comment (rcomer) [reply docs/src/whatsnew/latest.rst]: done

**Comment by rcomer** [comment]
Thanks @wjbenfold 😀

**Pull request merged**